flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartBackoffTimeStrategyFactoryLoader.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.RestartStrategyOptions;
+import org.apache.flink.runtime.executiongraph.restart.NoOrFixedIfCheckpointingEnabledRestartStrategyFactory;
+import org.apache.flink.runtime.executiongraph.restart.RestartStrategy;
+
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A utility class to load {@link RestartBackoffTimeStrategy.Factory} from the configuration.
+ * It respects the configs for the legacy {@link RestartStrategy}.
+ */
+public final class RestartBackoffTimeStrategyFactoryLoader {
+
+	private RestartBackoffTimeStrategyFactoryLoader() {
+	}
+
+	/**
+	 * Creates {@link RestartBackoffTimeStrategy.Factory} from the given configuration.
+	 *
+	 * <p>The strategy factory is decided in order as follows:
+	 * <ol>
+	 *     <li>Strategy set within job graph, i.e. {@link RestartStrategies.RestartStrategyConfiguration},
+	 * unless the config is {@link RestartStrategies.FallbackRestartStrategyConfiguration}.</li>
+	 *     <li>Strategy set in the cluster(server-side) config (flink-conf.yaml),
+	 * unless the strategy is not specified</li>
+	 *     <li>{@link FixedDelayRestartBackoffTimeStrategy.FixedDelayRestartBackoffTimeStrategyFactory} if
+	 * checkpointing is enabled. Otherwise {@link NoRestartBackoffTimeStrategy.NoRestartBackoffTimeStrategyFactory}</li>
+	 * </ol>
+	 *
+	 * @param jobRestartStrategyConfiguration restart configuration given within the job graph
+	 * @param clusterConfiguration cluster(server-side) configuration
+	 * @param isCheckpointingEnabled if checkpointing is enabled for the job
+	 * @return new version restart strategy factory
+	 */
+	public static RestartBackoffTimeStrategy.Factory createRestartStrategyFactory(
+			final RestartStrategies.RestartStrategyConfiguration jobRestartStrategyConfiguration,
+			final Configuration clusterConfiguration,
+			final boolean isCheckpointingEnabled) {
+
+		checkNotNull(jobRestartStrategyConfiguration);
+		checkNotNull(clusterConfiguration);
+
+		return getJobRestartStrategyFactory(jobRestartStrategyConfiguration)
+			.orElse(getClusterRestartStrategyFactory(clusterConfiguration)
+				.orElse(getDefaultRestartStrategyFactory(isCheckpointingEnabled)));
+	}
+
+	private static Optional<RestartBackoffTimeStrategy.Factory> getJobRestartStrategyFactory(
+			final RestartStrategies.RestartStrategyConfiguration restartStrategyConfiguration) {
+
+		if (restartStrategyConfiguration instanceof RestartStrategies.NoRestartStrategyConfiguration) {
+			return Optional.of(NoRestartBackoffTimeStrategy.NoRestartBackoffTimeStrategyFactory.INSTANCE);
+		} else if (restartStrategyConfiguration instanceof RestartStrategies.FixedDelayRestartStrategyConfiguration) {
+			final RestartStrategies.FixedDelayRestartStrategyConfiguration fixedDelayConfig =
```

Comment:
I think this branch and the one before is not tested.

Reply:
Done.